search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt

```diff
@@ -0,0 +1,144 @@
+                key == JSONLD_TYPE_TERM || key == NGSILD_SCOPE_TERM ->
+                    mergeTypeOrScope(localValue, value)
+                key == NGSILD_CREATED_AT_TERM ->
+                    if ((value as String?).isBefore(localValue as String?)) value
+                    else localValue
+                key == NGSILD_MODIFIED_AT_TERM ->
+                    if ((localValue as String?).isBefore(value as String?)) value
+                    else localValue
+                else -> mergeAttribute(
+                    localValue,
+                    value,
+                    csr
+                ).bind()
+            }
+        }
+    }
+
+    fun mergeTypeOrScope(
+        currentValue: Any, // String || List<String> || Set<String>
+        remoteValue: Any
+    ) = when {
+        currentValue == remoteValue -> currentValue
+        currentValue is List<*> && remoteValue is List<*> -> (currentValue.toSet() + remoteValue.toSet()).toList()
+        currentValue is List<*> -> (currentValue.toSet() + remoteValue).toList()
+        remoteValue is List<*> -> (remoteValue.toSet() + currentValue).toList()
+        else -> listOf(currentValue, remoteValue)
+    }
+
+    /**
+     * Implements 4.5.5 - Multi-Attribute Support
+     */
+    fun mergeAttribute(
+        currentAttribute: Any,
+        remoteAttribute: Any,
+        csr: ContextSourceRegistration
+    ): Either<NGSILDWarning, Any> = either {
+        val currentInstances = groupInstancesByDataSetId(currentAttribute, csr).bind().toMutableMap()
+        val remoteInstances = groupInstancesByDataSetId(remoteAttribute, csr).bind()
+        remoteInstances.entries.forEach { (datasetId, remoteInstance) ->
+            val currentInstance = currentInstances[datasetId]
+            when {
+                currentInstance == null -> currentInstances[datasetId] = remoteInstance
+                csr.isAuxiliary() -> {}
+                currentInstance.isBefore(remoteInstance, NGSILD_OBSERVED_AT_TERM) ->
+                    currentInstances[datasetId] = remoteInstance
+                remoteInstance.isBefore(currentInstance, NGSILD_OBSERVED_AT_TERM) -> {}
+                currentInstance.isBefore(remoteInstance, NGSILD_MODIFIED_AT_TERM) ->
+                    currentInstances[datasetId] = remoteInstance
+                remoteInstance.isBefore(currentInstance, NGSILD_MODIFIED_AT_TERM) -> {}
+                // if there is no discriminating factor choose one at random
+                nextBoolean() -> currentInstances[datasetId] = remoteInstance
+                else -> {}
+            }
+        }
+        val values = currentInstances.values.toList()
+        if (values.size == 1) values[0] else values
+    }
+
+    // do not work with CORE MEMBER since they are nor list nor map
+    private fun groupInstancesByDataSetId(
+        attribute: Any,
+        csr: ContextSourceRegistration
+    ): Either<NGSILDWarning, AttributeByDataSetId> =
+        when (attribute) {
```

Comment:
note that this will not work if entity is asked for using the simplified representation (and what about the GeoJSON representation?). or we always ask for the normalized representation to context sources and the final representation is done in the main context broker (as is currently done in `EntityHandler`)? (in this case, we should take care to not transmit the request parameter to the context source).

Reply:
If we query the CSR for the simplified representation (or geojson) we don't get the observedAt information so we can't merge the data properly. So i think the final representation shall be in the main broker. ( and i will remove the request param from the CSRrequest)